feat: add numtide devshell integration

oddlama opened the pull request:
This add integration with the numtide devshell, see #158 for more information

oddlama commented on src/implementation.nix:
```diff
@@ -278,6 +278,42 @@ in {
+            crate: cfg:
+            l.optional (cfg.numtideDevshell != null) {
+              ${cfg.numtideDevshell} = {
+                packagesFrom = [ config.nci.toolchains.shell ];
```
I'm not sure whether this is good style. If two crates would specify the same `numtideDevshell`, this would get added twice. It's not a problem in practice, although I'm not sure what this package entails. I'm open to suggestions if you consider this is a problem.

yusdacra replied:
it's just the rust toolchain with devshell specific additions added (think rustfmt, clippy, rust-src etc), the build toolchain doesnt have those to keep the derivation closure smaller. it should be okay if there are two of these, i don't think it would cause a problem in practice (nix should just realize they are the same anyway), so we can just keep it

yusdacra commented:
btw i see that it only generates a devshell for a crate, it would be nice if it also generated a devshell for a whole project aswell (if specified), to keep things consistent with how nci does it

oddlama commented:
Good point, I've also added some other stuff:
- Allow `numtideDevshell` option on both projects and crates
- Declare a phantom option `devshells` that silently merges as a no-op with the upstream option from numtide/devshell. This is necessary to retain compatibility with projects not using numtide/devshell without requiring it to be pulled in as a dependency.
- Added an example to showcase the integration.

Let me know if you have any further suggestions.

yusdacra commented:
thanks!
